pkg/commands/helpers/kubectl_auth.go

```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		return fmt.Errorf("error writing to stdout: %v", err)
+	}
+
+	if !isCached {
+		if err := os.MkdirAll(filepath.Dir(cacheFilePath), 0755); err != nil {
+			klog.Warningf("failed to make cache directory for %q: %v", cacheFilePath, err)
+		}
+		if err := ioutil.WriteFile(cacheFilePath, b, 0600); err != nil {
+			klog.Warningf("failed to write cache file %q: %v", cacheFilePath, err)
+		}
+	}
+
+	return nil
+}
+
+// ExecCredential specifies the client.authentication.k8s.io ExecCredential object
+type ExecCredential struct {
+	APIVersion string               `json:"apiVersion,omitempty"`
+	Kind       string               `json:"kind,omitempty"`
+	Status     ExecCredentialStatus `json:"status"`
+}
+
+// ExecCredentialStatus specifies the status of the client.authentication.k8s.io ExecCredential object
+type ExecCredentialStatus struct {
+	ClientCertificateData string    `json:"clientCertificateData,omitempty"`
+	ClientKeyData         string    `json:"clientKeyData,omitempty"`
+	ExpirationTimestamp   time.Time `json:"expirationTimestamp,omitempty"`
+}
+
+func cacheFilePath(kopsStateStore string, clusterName string) string {
+	var b bytes.Buffer
+	b.WriteString(kopsStateStore)
+	b.WriteByte(0)
+	b.WriteString(clusterName)
+	b.WriteByte(0)
+
+	hash := fmt.Sprintf("%x", sha256.New().Sum(b.Bytes()))
+	sanitizedName := strings.Map(func(r rune) rune {
+		switch {
+		case r >= 'a' && r <= 'z':
+			return r
+		case r >= 'A' && r <= 'Z':
+			return r
+		case r >= '0' && r <= '9':
+			return r
+		default:
+			return '_'
+		}
+	}, clusterName)
+	return filepath.Join(homedir.HomeDir(), ".kube", "cache", "kops-authentication", sanitizedName+"_"+hash)
+}
+
+func loadCachedExecCredential(cacheFilePath string) (*ExecCredential, error) {
+	b, err := ioutil.ReadFile(cacheFilePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// expected - a cache miss
+			return nil, nil
+		} else {
+			return nil, err
+		}
+	}
+
+	execCredential := &ExecCredential{}
+	if err := json.Unmarshal(b, execCredential); err != nil {
+		return nil, fmt.Errorf("error parsing: %v", err)
+	}
+
+	if execCredential.Status.ExpirationTimestamp.Before(time.Now()) {
+		return nil, nil
+	}
+
+	if execCredential.Status.ClientCertificateData == "" || execCredential.Status.ClientKeyData == "" {
+		return nil, fmt.Errorf("no credentials in cached file")
+	}
+
+	return execCredential, nil
+}
+
+func buildCredentials(ctx context.Context, f *util.Factory, options *HelperKubectlAuthOptions) (*ExecCredentialStatus, error) {
+	clientset, err := f.Clientset()
+	if err != nil {
+		return nil, err
+	}
+
+	cluster, err := clientset.GetCluster(ctx, options.ClusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	if cluster == nil {
+		return nil, fmt.Errorf("cluster not found %q", options.ClusterName)
+	}
+
+	keyStore, err := clientset.KeyStore(cluster)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get cluster keystore: %v", err)
+	}
+
+	req := pki.IssueCertRequest{
```

Comment:
This is based on #9593

Reply:
looks good. Might want to put what we think the userid is in the `CommonName`.

Reply:
I'll try to use user.Current() (which I believe is cgo safe now), doing `kubecfg-$name`  Happy to change to something better!